crates/sqlbuiltins/src/functions/table/read_blob.rs

```diff
@@ -0,0 +1,361 @@
+        let file_compression_type = self.file_compression_type;
+
+        Ok(Box::pin(async move {
+            let options = GetOptions::default();
+            let result = store.get_opts(file_meta.location(), options).await?;
+
+            // manually add columns based on the projected schema
+            let mut columns = Vec::new();
+
+            let mut size = schema
+                .column_with_name("size")
+                .map(|_| Arc::new(Int64Array::from(vec![result.meta.size as i64])) as ArrayRef);
+
+            let mut last_modified = schema.column_with_name("last_modified").map(|_| {
+                Arc::new(TimestampNanosecondArray::from_vec(
+                    vec![result.meta.last_modified.timestamp_nanos()],
+                    None,
+                )) as ArrayRef
+            });
+
+            let mut filename = schema.column_with_name("filename").map(|_| {
+                Arc::new(StringArray::from(vec![result.meta.location.to_string()])) as ArrayRef
+            });
+
+            let mut content = match schema.column_with_name("content") {
+                Some(_) => {
+                    let len = result.range.end - result.range.start;
+                    match result.payload {
+                        object_store::GetResultPayload::File(mut file, _) => {
+                            let mut bytes = match file_meta.range {
+                                None => file_compression_type.convert_read(file)?,
+                                Some(_) => {
+                                    file.seek(SeekFrom::Start(result.range.start as _))?;
+                                    let limit = result.range.end - result.range.start;
+                                    file_compression_type.convert_read(file.take(limit as u64))?
+                                }
+                            };
+                            let mut data = Vec::new();
+                            bytes.read_to_end(&mut data)?;
+
+                            Some(Arc::new(BinaryArray::from_vec(vec![&data])) as ArrayRef)
+                        }
+                        object_store::GetResultPayload::Stream(s) => {
+                            let s = s.map_err(DataFusionError::from);
+
+                            let s = file_compression_type.convert_stream(s.boxed())?.fuse();
+                            let bytes = collect_bytes(s, Some(len)).await?;
+                            Some(Arc::new(BinaryArray::from_vec(vec![&bytes])) as ArrayRef)
+                        }
+                    }
+                }
+                None => None,
+            };
+
+            for field in schema.fields() {
+                // we need to do this hacky option take because of the borrow checker.
+                // these matches should only ever match once, but the borrow checker doesn't know that.
+                // so we just wrap the value in an option to ensure it's only taken once.
```

Comment:
Instead of doing this, can we simply create the columns at the end? This code feels a bit hard to read. We could just have all the variables and then create different columns (and eventually record batch) from them at the end.

Reply:
@vrongmeal I refactored this to instead build the columns with an index, then reorder them by the index. I think this is a lot more readable.